packages/fetchai/connections/gym/connection.py

```diff
@@ -118,29 +113,38 @@ def handle_gym_message(self, envelope: Envelope) -> None:
                 protocol_id=GymMessage.protocol_id,
                 message=msg,
             )
-            self._send(envelope)
+            await self._send(envelope)
         elif gym_message.performative == GymMessage.Performative.RESET:
-            self.gym_env.reset()  # type: ignore
+            await self._run_in_executor(self.gym_env.reset)
         elif gym_message.performative == GymMessage.Performative.CLOSE:
-            self.gym_env.close()  # type: ignore
+            await self._run_in_executor(self.gym_env.close)
 
-    def _send(self, envelope: Envelope) -> None:
+    async def _send(self, envelope: Envelope) -> None:
         """Send a message.
 
         :param envelope: the envelope
         :return: None
         """
-        destination = envelope.to
-        self._queues[destination].put_nowait(envelope)
+        if self._queue is None:  # pragma: nocover
+            raise ValueError("Channel is not connected")
+        assert envelope.to == self.address, "Invalid destination address"
+        await self._queue.put(envelope)
 
-    def disconnect(self) -> None:
+    async def disconnect(self) -> None:
         """
         Disconnect.
 
         :return: None
         """
-        with self._lock:
-            self._queues.pop(self.address, None)
+        if self._queue is not None:
+            await self._queue.put(None)
+            self._queue = None
+
+    async def get(self) -> Optional[Envelope]:
+        """Get incoming envelope."""
+        if self._queue is None:  # pragma: nocover
+            raise ValueError("Channel is not connected")
```

Comment:
If you wrap `self._queue` in a property you can reduce some duplication of code..